src/main/kotlin/org/rust/lang/refactoring/implementMembers/impl.kt

```diff
@@ -45,20 +47,63 @@ private fun findMembersToImplement(impl: RsImplItem): Pair<TraitImplementationIn
     return implInfo to trait
 }
 
-private fun insertNewTraitMembers(selected: Collection<RsNamedElement>, members: RsMembers, subst: Substitution) {
+private fun insertNewTraitMembers(selected: Collection<RsAbstractable>, existingMembers: RsMembers, trait: BoundElement<RsTraitItem>) {
     checkWriteAccessAllowed()
     if (selected.isEmpty()) return
 
-    val templateImpl = RsPsiFactory(members.project).createMembers(
-        selected.filterIsInstance<RsFunction>(),
-        selected.filterIsInstance<RsTypeAlias>(),
-        selected.filterIsInstance<RsConstant>(),
-        subst
-    )
-    val lastMethodOrBrace = members.functionList.lastOrNull() ?: members.lbrace ?: return
-    members.addRangeAfter(
-        templateImpl.lbrace.nextSibling,
-        templateImpl.rbrace?.prevSibling,
-        lastMethodOrBrace
-    )
+    val templateImpl = RsPsiFactory(existingMembers.project).createMembers(selected, trait.subst)
+
+    val traitMembers = trait.element.members!!.childrenOfType<RsAbstractable>()
+    val newMembers = templateImpl.childrenOfType<RsAbstractable>()
+
+    // [1] First, check if the order of the existingMembers already implemented
+    // matches the order of existingMembers in the trait declaration.
+    val existingMembersOrder = existingMembers.childrenOfType<RsAbstractable>().map {
+        existingMember -> traitMembers.indexOfFirst { it.name == existingMember.name }
+    }
+    val areExistingMembersInTheRightOrder = existingMembersOrder == existingMembersOrder.sorted()
+
+    for ((index, newMember) in newMembers.withIndex()) {
+        val posInTrait = traitMembers.indexOfFirst { it.name == newMember.name }
+
+        // If [1] does not hold, the first new member we will append at the end of the implementation.
+        // All the other ones will consequently be inserted at the right position in relation to that very first one.
+        val anchor = if (!areExistingMembersInTheRightOrder && index == 0) {
+            existingMembers.childrenOfType<RsAbstractable>().lastOrNull()
+        } else {
+            existingMembers.childrenOfType<RsAbstractable>().findLast {
+                existingMember -> traitMembers.indexOfFirst { it.name == existingMember.name } < posInTrait
+            }
+        } ?: existingMembers.lbrace ?: return
+
+        // If the newly added item is a function, we add an extra line between it and each of its siblings.
+        val addedMember = existingMembers.addAfter(newMember, anchor)
+
+        val previousAbstractable = addedMember.leftSiblings.find { it is RsAbstractable }
+        if (previousAbstractable != null) {
```

Comment:
check for null is redundant here

Reply:
`previousAbstractable` will be null when inserting a stub implementation of a member at the very first position in the `impl`. In fact, when I remove this check, I get 10 test failures.